sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -395,11 +395,24 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
   }
 
   /** Appends the string represent of this node and its children to the given StringBuilder. */
-  protected def generateTreeString(depth: Int, builder: StringBuilder): StringBuilder = {
-    builder.append(" " * depth)
-    builder.append(simpleString)
-    builder.append("\n")
-    children.foreach(_.generateTreeString(depth + 1, builder))
+  protected def generateTreeString(
+      depth: Int, lastChildren: Seq[Boolean], builder: StringBuilder): StringBuilder = {
+    val prefix = if (depth == 0) {
+      ""
+    } else {
+      (lastChildren.init.map { isLast =>
+        if (isLast) "   " else ":  "
+      } :+ (if (lastChildren.last) s"+- " else ":- ")).mkString
```

Comment:
this may be arguably more readable:
```
lastChildren.init.map { isLast =>
  if (isLast) "   " else ":  "
}.mkString + (if (lastChildren.last) "+- " else ":- ")
```